badger/badger.go

```diff
@@ -0,0 +1,346 @@
+	iterators []*badger.Iterator
+	mutex     sync.RWMutex
+	store     *Store
+	badgerTx  *badger.Txn
+}
+
+func (b *tx) Unwrap() interface{} {
+	return b.badgerTx
+}
+
+func (b *tx) GetShelfReader(shelfName string) stoabs.Reader {
+	return b.getBucket(shelfName)
+}
+
+func (b *tx) GetShelfWriter(shelfName string) (stoabs.Writer, error) {
+	return &badgerShelf{name: shelfName, tx: b}, nil
+}
+
+func (b *tx) getBucket(shelfName string) stoabs.Reader {
+	return &badgerShelf{name: shelfName, tx: b}
+}
+
+func (b *tx) Store() stoabs.KVStore {
+	return b.store
+}
+
+func (b *tx) newIterator() *badger.Iterator {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+
+	iterator := b.badgerTx.NewIterator(badger.DefaultIteratorOptions)
+	b.iterators = append(b.iterators, iterator)
+
+	return iterator
+}
+
+func (b *tx) rollback() {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+
+	for _, it := range b.iterators {
+		it.Close()
+	}
+	b.badgerTx.Discard()
+}
+
+func (b *tx) commit() error {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+
+	for _, it := range b.iterators {
+		it.Close()
+	}
+	return b.badgerTx.Commit()
+}
+
+type badgerShelf struct {
+	name string
+	tx   *tx
+}
+
+func (t badgerShelf) key(key stoabs.Key) stoabs.Key {
+	myBytes := []byte(t.name)
+	newKey := stoabs.BytesKey(append(myBytes, key.Bytes()...))
+	return newKey
+}
+
+func (t badgerShelf) Get(key stoabs.Key) ([]byte, error) {
+	item, err := t.tx.badgerTx.Get(t.key(key).Bytes())
+	if err != nil {
+		if errors.Is(err, badger.ErrKeyNotFound) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	value := make([]byte, item.ValueSize())
+	item.ValueCopy(value)
+	return value, nil
+}
+
+func (t badgerShelf) Put(key stoabs.Key, value []byte) error {
+	return t.tx.badgerTx.Set(t.key(key).Bytes(), value)
+}
+
+func (t badgerShelf) Delete(key stoabs.Key) error {
+	return t.tx.badgerTx.Delete(t.key(key).Bytes())
+}
+
+// Stats are currently broken
+func (t badgerShelf) Stats() stoabs.ShelfStats {
+	var onDiskSize, keyCount uint
+	tables := t.tx.store.DB.Tables()
+	prefix := []byte(t.name)
+	for _, ti := range tables {
+		if bytes.HasPrefix(ti.Left, prefix) && bytes.HasPrefix(ti.Right, prefix) {
+			onDiskSize += uint(ti.OnDiskSize)
+			keyCount += uint(ti.KeyCount)
+			//uncompressedSize += uint64(ti.UncompressedSize)
+		}
+	}
+	return stoabs.ShelfStats{
+		NumEntries: keyCount,
+		ShelfSize:  onDiskSize,
+	}
+}
+
+func (t badgerShelf) Iterate(callback stoabs.CallerFn, keyType stoabs.Key) error {
+	// closed by commit or rollback
+	it := t.tx.newIterator()
+	t.tx.mutex.RLock()
+	defer t.tx.mutex.RUnlock()
+
+	prefix := []byte(t.name)
+	for it.Seek(prefix); it.ValidForPrefix(prefix) && t.tx.ctx.Err() == nil; it.Next() {
+		item := it.Item()
+		k := item.Key()
+		if err := item.Value(func(v []byte) error {
+			kt, err := keyType.FromBytes(k[len(prefix):])
+			if err != nil {
+				return err
+			}
+			return callback(kt, v)
+		}); err != nil {
+			return err
+		}
+	}
+	if t.tx.ctx.Err() != nil {
```

Comment:
calling friends of friends isn't recommended

Reply:
placed ctx in shelf as well